### PR TITLE
Added hive connection retry properties

### PR DESCRIPTION
**Pull request by Yaliang**
Added hive metastore client connection creation retry driver properties.

- Added `HiveClientRetryConfig` class for all settings about the connection retry.
- Changed `CachingHiveMetastore.retry()` as `protected` to allow the added subclass `CachingHiveMetastoreRetryConfigurable` to override this method. `CachingHiveMetastoreRetryConfigurable` extends `CachingHiveMetastore` with the added retry driver properties.
- Wrapped part of `configuration` method (which is binding `HiveMetastore`) as a new `protected` method `bindMetastore` in `HiveClientModule` to allow the added subclass `HiveClientModuleRetryConfigurable` to bind `HiveMetastore` to `CachingHiveMetastoreRetryConfigurable`, instead of `CachingHiveMetastore`, by overriding the method `bindMetastore`. 
- Replaced `HiveClientModule` by `HiveClientModuleRetryConfigurable` in `HiveConnectorFactory`

### CODE REVIEW

**Comment by billonahill** [comment]
Is it possible to try to encapsulate as much of our custom edits into classes in the `twitter` package and then make minimal changes (i.e., only a few lines) to the code on master? This helps reviewing diffs made on master verses Twitter-specific patches. If we get too many Twitter-specific changes on top of the core codebase, we'll have a maintainability headache keeping consistent with changes to master and we'll risk becoming a forked version of the codebase. That was the intent of that package.

Also can we contribute these changes upstream to the presto master?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 1]
```diff
@@ -77,14 +77,7 @@ public void configure(Binder binder)
...
 
-        if (metastore != null) {
-            binder.bind(HiveMetastore.class).toInstance(metastore);
-        }
-        else {
-            binder.bind(HiveMetastore.class).to(CachingHiveMetastore.class).in(Scopes.SINGLETON);
-            newExporter(binder).export(HiveMetastore.class)
-                    .as(generatedNameOf(CachingHiveMetastore.class, connectorId));
-        }
+        bindMetastore(binder);
```
Comment (billonahill) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java]: Why is this change necessary? The method does the same thing and I don't see you overriding it. If you just need to swap the CachingHiveMetastore impl for another one, in this cases I'd say better to make the 2 line change to this class to do so.

File: presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.concurrent.ExecutorService;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+@ThreadSafe
+public class CachingHiveMetastoreRetryConfigurable
```
Comment (billonahill) [inline presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java]: please add javadocs explaining why this class is necessary. if this is just temporary for troubleshooting, let's say that.

File: presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * CachingHiveMetastoreRetryConfigurable is a subclass of CachingHiveMetastore.
+ * It added retry configuration and overrided the RetryDriver setup method
+ * retry().
+ * The class added the flexibility to set up retry configuration from a
+ * runtime injector in order to solve "Failed to connect hive metastore" issue.
```
Comment (billonahill) [inline presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java]: In discussions you mentioned that this code is only being added to debug this issue, implying that it shouldn't be permanently in place. Is that correct, or do we want it in the code even after that issues is resolved? If this is a temporary thing, please include that in the comments with a link to an issue to later remove it.

File: presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java [hunk 4]
```diff
@@ -35,11 +35,15 @@
...
+ * runtime injector in order to solve "Failed connecting to hive metastore" issue,
+ * where the local cache of metastore failed to create a connection to remote metastore
+ * via HiveCluster.
  * All retry configurations are propergated from a
  * HiveClientRetryConfig instance.
  *
- * @see IQ-221
+ * This fix is a temporary fix and will keep investigating the issue.
+ *
+ * @see <a href="https://jira.twitter.biz/browse/IQ-221">IQ-221</a>
```
Comment (billonahill) [inline presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java]: Sorry, what I meant was please include a link to a jira who's sole purpose is the remove of this code, which is blocked on IQ-221, if this code is in fact temporary. That way once IQ-221 is resolved, we still have a clean-up jira. Also, best to include only the bug name withiout the full path to jira, since this will be visible externally.

Comment (Yaliang) [inline presto-hive/src/main/java/com/facebook/presto/twitter/hive/CachingHiveMetastoreRetryConfigurable.java]: Oh, I see. Will add it.

**Comment by billonahill** [comment]
As discussed, I think we should hold off on shipping this until after we've learned the cause of the connection failures and attempted a fix. This change could potentially mask a real connection issue that should be corrected.

**Comment by Yaliang** [comment]
@billonahill Actually this is blocking us to test the presto query boundary. As we are still logging all information and already found a way to reproduce the error in devel, I wish we can ship this quick fix.  Even if we just add a new catalog with the retry properties for testing query boundary only and keep the hive_prod catalog throwing "Failed connection error" until found a real connection issue.

**Comment by billonahill** [comment]
@Yaliang I'm still somewhat concerned though that this patch could still have a detrimental impact on performance and that we should prioritize the true fix over progress with the use cases testing. When adding distributed retries to an already failing connection (potentially due to resources being exceeded) there is a potential to further destabilize the system. From your investigation so far, why do you think that will not happen in this case?

Also when a query fails, it it because a few workers exhibit the hive connection failure, or because many do? Does this failure happen across multiple queries on a given worker when it happens, or does it isolate itself to the query being processed?

File: pom.xml [hunk 5]
```diff
@@ -10,7 +10,7 @@
 
     <groupId>com.facebook.presto</groupId>
     <artifactId>presto-root</artifactId>
-    <version>0.149-tw-0.27</version>
+    <version>0.149-tw-0.28</version>
```
Comment (billonahill) [inline pom.xml]: This PR should not add functionality and rev the presto version at the same time. We should rev versions in PRs of dedicated to just that.

Comment (Yaliang) [inline pom.xml]: True. Should not include the reversion. 
Pushed with a correction of the typo. Will remove it.

**Comment by Yaliang** [comment]
@billonahill The impact on performance caused by retries would be limited because there is a max thread pool limitation for hive metastore executor. Based on the current investigation, only the coordinator connects to hive metastore, each catalog has a local cache and have a corresponding thread pool. I believe it will not be worse than just reduce the max thread pool size to somewhere not causing hive connection failure.

Luckily, we are able to see the overall 'inuse' socket number with our individual role in the presto coordinator process. In devel, the failure happens when socket number steps up for 100~200 in sudden. And in prod, the failure happens when socket number steps up for 6000~7000. These data are based on the reproducing approach in JIRA ticket(IQ-221). It is just a normal case in prod, that the socket number steps thousands up in sudden without hive connection failed. Based on this observation, I guess potentially exceeded resource is isolated with other network resources like the connection to worker nodes in some way.

When hive connection failure happens, the error can cross multiple queries, if they are sharing the same catalog. They share the same thread pool, the same cache of metastore, and share same error as well. It may become another story if the queries are in two different catalogs. I will try if the error will cross catalogs.

**Comment by Yaliang** [comment]
As the root cause has found and fixed in other patch, we can close this PR.
